R/persist-dm.R

```diff
@@ -0,0 +1,250 @@
+# @export
+dm_upsert <- function(target_dm, dm, ..., dry_run = FALSE) {
+  check_dots_empty()
+
+  dm_persist(target_dm, dm, tbl_upsert, top_down = TRUE, dry_run)
+}
+
+# dm_delete
+#
+# `dm_delete()` removes matching records, based on the primary key.
+# The order in which the tables are processed is reversed.
+#
+# @rdname persist-dm
+# @export
+dm_delete <- function(target_dm, dm, ..., dry_run = FALSE) {
+  check_dots_empty()
+
+  dm_persist(target_dm, dm, tbl_delete, top_down = FALSE, dry_run)
+}
+
+# dm_truncate
+#
+# `dm_truncate()` removes all records, only for tables in `dm`.
+# The order in which the tables are processed is reversed.
+#
+# @rdname persist-dm
+# @export
+dm_truncate <- function(target_dm, dm, ..., dry_run = FALSE) {
+  check_dots_empty()
+
+  dm_persist(target_dm, dm, tbl_truncate, top_down = FALSE, dry_run)
+}
+
+dm_persist <- function(target_dm, dm, operation, top_down, dry_run = FALSE) {
+  dm_check_persist(target_dm, dm)
+
+  dm_run_persist(target_dm, dm, operation, top_down, dry_run)
+}
+
+dm_check_persist <- function(target_dm, dm) {
+  check_not_zoomed(target_dm)
+  check_not_zoomed(dm)
+
+  check_same_src(target_dm, dm)
+  check_tables_superset(target_dm, dm)
+  tables <- dm_get_tables(dm)
+  walk2(dm_get_tables(target_dm)[names(tables)], tables, check_columns_superset)
+  check_keys_compatible(target_dm, dm)
+}
+
+check_same_src <- function(target_dm, dm) {
+  tables <- c(dm_get_tables_impl(target_dm), dm_get_tables_impl(dm))
+  if (!all_same_source(tables)) {
+    abort_not_same_src()
+  }
+}
+
+check_tables_superset <- function(target_dm, dm) {
+  tables_missing <- setdiff(src_tbls_impl(dm), src_tbls_impl(target_dm))
+  if (has_length(tables_missing)) {
+    abort_tables_missing(tables_missing)
+  }
+}
+
+check_columns_superset <- function(target_tbl, tbl) {
+  columns_missing <- setdiff(colnames(tbl), colnames(target_tbl))
+  if (has_length(columns_missing)) {
+    abort_columns_missing(columns_missing)
+  }
+}
+
+check_keys_compatible <- function(target_dm, dm) {
+  # FIXME
+}
+
+
+
+dm_run_persist <- function(target_dm, dm, tbl_op, top_down, dry_run) {
+  # topologically sort tables
+  graph <- dm::create_graph_from_dm(target_dm, directed = TRUE)
+  topo <- igraph::topo_sort(graph, mode = if (top_down) "in" else "out")
+  tables <- intersect(names(topo), src_tbls(dm))
+
+  # extract keys
+  target_tbls <- dm_get_tables_impl(target_dm)[tables]
+  tbls <- dm_get_tables_impl(dm)[tables]
+
+  # FIXME: Extract keys for upsert and delete
+  # Use keyholder?
+
+  # run operation(target_tbl, source_tbl, dry_run = dry_run) for each table
+  op_results <- map2(target_tbls, tbls, tbl_op, dry_run = dry_run)
+
+  # operation() returns NULL if no table is needed, otherwise a tbl
+  new_tables <- compact(op_results)
+
+  out <-
+    target_dm %>%
+    dm_patch_tbl(!!!new_tables)
+
+  if (dry_run) {
+    out
+  } else {
+    invisible(out)
+  }
+}
+
+dm_patch_tbl <- function(dm, ...) {
+  check_not_zoomed(dm)
```

Comment:
One more thing: So far this function is only called in `dm_run_persist()` which is called by `dm_persist()`.
In `dm_persist()` all sorts of checks are done, also the `check_not_zoomed()` one.
So unless you are planning to use this function also in other places, we could remove this line
```suggestion
```